File: backend/app/schemas/gameplay.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.certificate import CertificateRead
from app.schemas.reward import BranchResolutionRead, EarnedRewardRead


class PlayableGameRead(BaseModel):
    id: int
    title: str
    description: str
    category: str
    creator_id: int
    level_count: int
    status: str


class PlayerProgressRead(BaseModel):
    id: int
    player_id: int
    game_id: int
    current_level_id: int | None
    status: str
    completed_level_ids: list[int]
    last_score: int | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PlayableQuestionRead(BaseModel):
    id: str
    prompt: str
    options: list[dict]
    xp: int


class PlayableLevelRead(BaseModel):
    id: int
    game_id: int
    sequence_number: int
    title: str
    learning_objective: str
    topic_explanation: str
    difficulty: str
    scenarios: list[dict]
    questions: list[PlayableQuestionRead]
    reward: dict
    pass_score: int
    branching_suggestion: dict


class GameplayStartRead(BaseModel):
    progress: PlayerProgressRead
    level: PlayableLevelRead


class LevelAttemptRead(BaseModel):
    id: int
    player_id: int
    game_id: int
    level_id: int
    status: str
    score: int | None
    started_at: datetime
    completed_at: datetime | None

    model_config = {"from_attributes": True}


class StartAttemptRead(BaseModel):
    attempt: LevelAttemptRead
    level: PlayableLevelRead


class SubmitAnswerRequest(BaseModel):
    question_id: str
    selected_answer: str


class SubmitAnswerRead(BaseModel):
    question_id: str
    selected_answer: str
    is_correct: bool
    feedback: dict


class CompleteAttemptRead(BaseModel):
    attempt: LevelAttemptRead
    score: int
    passed: bool
    reward: EarnedRewardRead | None
    branch: BranchResolutionRead
    progress: PlayerProgressRead
    certificate: CertificateRead | None = None

File: frontend/src/api/gameplay.ts
import { API_BASE_URL } from './config';
import { Certificate } from './certificates';
import { BranchResolution, EarnedReward } from './rewards';

export type PlayableGame = {
  id: number;
  title: string;
  description: string;
  category: string;
  creator_id: number;
  level_count: number;
  status: string;
};

export type PlayerProgress = {
  id: number;
  player_id: number;
  game_id: number;
  current_level_id: number | null;
  status: string;
  completed_level_ids: number[];
  last_score: number | null;
  created_at: string;
  updated_at: string;
};

export type PlayableQuestion = {
  id: string;
  prompt: string;
  options: Array<Record<string, unknown>>;
  xp: number;
};

export type PlayableLevel = {
  id: number;
  game_id: number;
  sequence_number: number;
  title: string;
  learning_objective: string;
  topic_explanation: string;
  difficulty: string;
  scenarios: Array<Record<string, unknown>>;
  questions: PlayableQuestion[];
  reward: Record<string, unknown>;
  pass_score: number;
  branching_suggestion: Record<string, unknown>;
};

export type GameplaySession = {
  progress: PlayerProgress;
  level: PlayableLevel;
};

export type LevelAttempt = {
  id: number;
  player_id: number;
  game_id: number;
  level_id: number;
  status: string;
  score: number | null;
  started_at: string;
  completed_at: string | null;
};

export type StartAttempt = {
  attempt: LevelAttempt;
  level: PlayableLevel;
};

export type AnswerFeedback = {
  question_id: string;
  selected_answer: string;
  is_correct: boolean;
  feedback: Record<string, unknown>;
};

export type CompleteAttempt = {
  attempt: LevelAttempt;
  score: number;
  passed: boolean;
  reward: EarnedReward | null;
  branch: BranchResolution;
  progress: PlayerProgress;
  certificate: Certificate | null;
};

export async function listPlayableGames(token: string): Promise<PlayableGame[]> {
  return request<PlayableGame[]>('/api/player/games', { token });
}

export async function startGame(token: string, gameId: number): Promise<GameplaySession> {
  return request<GameplaySession>(`/api/player/games/${gameId}/start`, { method: 'POST', token });
}

export async function getGameProgress(token: string, gameId: number): Promise<GameplaySession> {
  return request<GameplaySession>(`/api/player/games/${gameId}/progress`, { token });
}

export async function startLevelAttempt(token: string, levelId: number): Promise<StartAttempt> {
  return request<StartAttempt>(`/api/player/levels/${levelId}/attempts`, { method: 'POST', token });
}

export async function submitAnswer(
  token: string,
  attemptId: number,
  questionId: string,
  selectedAnswer: string,
): Promise<AnswerFeedback> {
  return request<AnswerFeedback>(`/api/player/attempts/${attemptId}/answers`, {
    method: 'POST',
    token,
    body: JSON.stringify({ question_id: questionId, selected_answer: selectedAnswer }),
  });
}

export async function completeAttempt(token: string, attemptId: number): Promise<CompleteAttempt> {
  return request<CompleteAttempt>(`/api/player/attempts/${attemptId}/complete`, { method: 'POST', token });
}

async function request<T>(path: string, options: RequestInit & { token: string }): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${options.token}`,
      ...options.headers,
    },
  });

  if (!response.ok) {
    let message = 'Something went wrong.';
    try {
      const payload = await response.json();
      message = payload.detail ?? message;
    } catch {
      message = response.statusText || message;
    }
    throw new Error(message);
  }
  return response.json() as Promise<T>;
}
